File: library/src/main/java/io/vithor/yamvpframework/mvp/presenter/SimplePresenter.java
package io.vithor.yamvpframework.mvp.presenter;

import android.support.annotation.NonNull;

import com.orhanobut.logger.Logger;

import org.jetbrains.annotations.NotNull;

import io.vithor.yamvpframework.ErrorContainer;
import io.vithor.yamvpframework.ResponseContainer;
import io.vithor.yamvpframework.mvp.presenter.sketch.TypedSketch;

/**
 * Created by Vithorio Polten on 5/5/15.
 */
public abstract class SimplePresenter<ModelType, SK extends TypedSketch<ModelType>, ResponseType> extends BasePresenter<SK> {

    @Override
    protected void onViewAttached() {
        loadData(PresenterAction.INITIAL_LOAD);
    }

    protected final PresenterCallback<ModelType, ResponseType> defaultHandler(PresenterAction action) {
        return new PresenterCallback<ModelType, ResponseType>(this, action) {
            @Override
            public void success(ModelType modelType, @NotNull ResponseContainer<ResponseType> response, @NotNull PresenterAction action) {
                successHandler(action, modelType, response);
            }
        };
    }

    protected abstract void loadData(PresenterAction action);

    protected void showLoading(PresenterAction action) {
        SK view = getView();
        if (view != null) {
            view.showLoading(action);
        }

    }

    protected void successHandler(PresenterAction action, ModelType t, ResponseContainer rawResponse) {
        SK view = getView();
        if (view != null) {
            view.showData(t);
            view.showContent();
        }
    }

    @Override
    public void handleRestFailure(@NonNull ErrorContainer error, @NonNull PresenterAction action) throws ViewDetachedException {
        Logger.e(error.getError(), "RestError");
//        RetrofitError retrofitError = (RetrofitError) error.getError();

//        switch (retrofitError.getKind()) {
//            case NETWORK:
//                getView().showNetworkError(action == PresenterAction.PULL_TO_REFRESH);
//                break;
//            case CONVERSION:
//                getView().showServerError(action == PresenterAction.PULL_TO_REFRESH);
//                break;
//            default:
//                getView().showServerError(action == PresenterAction.PULL_TO_REFRESH);
//        }
    }
}

File: mvp/src/main/java/io/vithor/yamvpframework/mvp/presenter/SimplePresenter.kt
package io.vithor.yamvpframework.mvp.presenter

import io.vithor.yamvpframework.mvp.ResponseContainer
import io.vithor.yamvpframework.mvp.presenter.sketch.TypedSketch

/**
 * Created by Vithorio Polten on 5/5/15.
 */
abstract class SimplePresenter<ModelType, SK : TypedSketch<ModelType>, ResponseType> : BasePresenter<SK>() {

    override fun onViewAttached() {
        loadData(PresenterAction.INITIAL_LOAD)
    }

    protected fun defaultHandler(action: PresenterAction): PresenterCallback<ModelType, ResponseType> {
        return object : PresenterCallback<ModelType, ResponseType>(this, action) {
            override fun success(t: ModelType, response: ResponseContainer<ResponseType>, action: PresenterAction) {
                successHandler(action, t, response)
            }
        }
    }

    protected abstract fun loadData(action: PresenterAction)

    protected fun showLoading(action: PresenterAction) {
        val view = view
        view?.showLoading(action)

    }

    protected fun successHandler(action: PresenterAction, t: ModelType, rawResponse: ResponseContainer<ResponseType>) {
        val view = view
        if (view != null) {
            view.showData(t)
            view.showContent()
        }
    }


    /*    @Override
    public void handleRestFailure(@NonNull ErrorContainer error, @NonNull PresenterAction action) throws ViewDetachedException {
        //Logger.e(error.getError(), "RestError");
//        RetrofitError retrofitError = (RetrofitError) error.getError();

//        switch (retrofitError.getKind()) {
//            case NETWORK:
//                getView().showNetworkError(action == PresenterAction.PULL_TO_REFRESH);
//                break;
//            case CONVERSION:
//                getView().showServerError(action == PresenterAction.PULL_TO_REFRESH);
//                break;
//            default:
//                getView().showServerError(action == PresenterAction.PULL_TO_REFRESH);
//        }
    }*/
}
